### PR TITLE
[fpmsyncd] Implement pending route suppression feature

### PR DESCRIPTION
**Pull request by stepanblyschak**
**What I did**

I implemented support to enable pending routes suppression feature. When this feature is enabled, fpmsyncd will wait for reply from orchagent before sending offload status message to zebra.

**Why I did it**

This is done to not announce routes which aren't yet offloaded in HW.

**How I verified it**

UT and manual tests.

**Details if related**

### CODE REVIEW

File: fpmsyncd/fpmlink.cpp [hunk 1]
```diff
@@ -277,3 +279,31 @@ void FpmLink::processFpmMessage(fpm_msg_hdr_t* hdr)
         nlmsg_free(msg);
     }
 }
+
+bool FpmLink::send(nlmsghdr* nl_hdr)
+{
+    fpm_msg_hdr_t hdr;
```
Comment (stepanblyschak) [inline fpmsyncd/fpmlink.cpp:285]:
```suggestion
    fpm_msg_hdr_t hdr{};
```

File: fpmsyncd/fpmlink.cpp [hunk 2]
```diff
@@ -277,3 +279,31 @@ void FpmLink::processFpmMessage(fpm_msg_hdr_t* hdr)
...
+
+    size_t len = fpm_msg_align(sizeof(hdr) + nl_hdr->nlmsg_len);
+
+    hdr.version = FPM_PROTO_VERSION;
+    hdr.msg_type = FPM_MSG_TYPE_NETLINK;
+    hdr.msg_len = htons(static_cast<uint16_t>(len));
+
+    memcpy(m_sendBuffer, &hdr, sizeof(hdr));
+    memcpy(m_sendBuffer + sizeof(hdr), nl_hdr, nl_hdr->nlmsg_len);
```
Comment (stepanblyschak) [inline fpmsyncd/fpmlink.cpp:294]: Do safe copy (with original buffer size check)

File: fpmsyncd/fpmsyncd.cpp [hunk 3]
```diff
@@ -57,11 +69,23 @@ int main(int argc, char **argv)
...
+        routeResponseChannel = std::make_unique<NotificationConsumer>(&applStateDb, routeResponseChannelName);
+        sync.setSuppressionEnabled(true);
+    }
+
     while (true)
     {
         try
         {
             FpmLink fpm(&sync);
+            sync.m_fpmInterface = &fpm;
```
Comment (stepanblyschak) [inline fpmsyncd/fpmsyncd.cpp:87]: Think about avoiding dependency loop

File: fpmsyncd/routesync.cpp [hunk 4]
```diff
@@ -1222,3 +1269,180 @@ string RouteSync::getNextHopWt(struct rtnl_route *route_obj)
 
     return result;
 }
+
+void RouteSync::sendOffloadReply(struct nlmsghdr* hdr)
+{
+    SWSS_LOG_ENTER();
+
+    assert(m_fpmInterface);
```
Comment (stepanblyschak) [inline fpmsyncd/routesync.cpp:1277]: Relates to m_fpmInterface <-> RouteSync dependency loop

File: fpmsyncd/routesync.cpp [hunk 5]
```diff
@@ -1222,3 +1269,180 @@ string RouteSync::getNextHopWt(struct rtnl_route *route_obj)
...
+}
+
+void RouteSync::sendOffloadReply(struct rtnl_route* route_obj)
+{
+    SWSS_LOG_ENTER();
+
+    nl_msg* msg{};
+    rtnl_route_build_add_request(route_obj, NLM_F_CREATE, &msg);
+
+    auto ownedMsg = std::unique_ptr<nl_msg, void(*)(nl_msg*)>(msg, nlmsg_free);
```
Comment (stepanblyschak) [inline fpmsyncd/routesync.cpp:1303]: ```makeUniqueWithDestructor```

File: fpmsyncd/routesync.cpp [hunk 6]
```diff
@@ -1222,3 +1269,180 @@ string RouteSync::getNextHopWt(struct rtnl_route *route_obj)
...
+    bool isSetOperation{false};
+    bool isSuccessReply{false};
+
+    if (!isSuppressionEnabled())
+    {
+        return;
+    }
+
+    auto colon = key.find(':');
+    if (colon != std::string::npos && key.substr(0, colon).find("Vrf") != std::string::npos)
```
Comment (stepanblyschak) [inline fpmsyncd/routesync.cpp:1332]: "Vrf" -> define?

File: fpmsyncd/routesync.cpp [hunk 7]
```diff
@@ -1222,3 +1269,180 @@ string RouteSync::getNextHopWt(struct rtnl_route *route_obj)
...
+        std::string field = fvField(fieldValue);
+        std::string value = fvValue(fieldValue);
+
+        if (field == "err_str")
+        {
+            isSuccessReply = (value == "SWSS_RC_SUCCESS");
+        }
+        else if (field == "protocol")
+        {
+            isSetOperation = true;
```
Comment (stepanblyschak) [inline fpmsyncd/routesync.cpp:1353]: Make a comment that proto is absent in case of DEL operation

File: fpmsyncd/routesync.cpp (hunk 6 above)
Comment (stepanblyschak) [reply fpmsyncd/routesync.cpp:1332]: There is no parsing of key from ROUTE_TABLE in fpmsyncd. It is present in orchagent where I have copied this logic from. But there is a macro for "Vrf"

File: fpmsyncd/fpmlink.cpp (hunk 2 above)
Comment (stepanblyschak) [reply fpmsyncd/fpmlink.cpp:294]: It's str* functions that have a safe variants. I just put a check before that len should be less then or eq m_bufSize.

**Review by volodymyrsamotiy — APPROVED** [review]

**Pull request closed without merging**